modules/core/src/main/java/org/apache/ignite/maintenance/MaintenanceRegistry.java

```diff
@@ -0,0 +1,132 @@
+ *             that requires user intervention)
+ *             or by user request (e.g. for a planned maintenance that requires
+ *             detaching node from the rest of the cluster).
+ *         </li>
+ *         <li>
+ *             Component responsible for handling this {@link MaintenanceTask}
+ *             on startup checks if the task is registered (thus it should go to Maintenance Mode).
+ *             If task is found component provides to {@link MaintenanceRegistry} its own implementation
+ *             of {@link MaintenanceWorkflowCallback} interface
+ *             via method {@link MaintenanceRegistry#registerWorkflowCallback(UUID, MaintenanceWorkflowCallback)}.
+ *         </li>
+ *         <li>
+ *             {@link MaintenanceWorkflowCallback} should provide {@link MaintenanceRegistry} with
+ *             {@link MaintenanceAction}s that are able to resolve maintenance task,
+ *             get information about it and so on.
+ *             Logic of these actions is completely up to the component providing it
+ *             and depends only on particular maintenance task.
+ *         </li>
+ *         <li>
+ *             When maintenance task is fixed, it should be removed from {@link MaintenanceRegistry}
+ *             with call {@link MaintenanceRegistry#unregisterMaintenanceTask(UUID)}.
+ *         </li>
+ *     </ul>
+ * </p>
+ */
+@IgniteExperimental
+public interface MaintenanceRegistry {
+    /**
+     * @return {@code True} if any maintenance task was found.
+     */
+    public boolean isMaintenanceMode();
+
+    /**
+     * @param task {@link MaintenanceTask} object with maintenance information that needs
+     *                                     to be stored to maintenance registry.
+     *
+     * @throws IgniteCheckedException If handling or storing maintenance task failed.
+     *
+     * @return Previously registered {@link MaintenanceTask} with the same ID
+     * or null if no tasks were registered for this ID.
+     */
+    public @Nullable MaintenanceTask registerMaintenanceTask(MaintenanceTask task) throws IgniteCheckedException;
+
+    /**
+     * Deletes {@link MaintenanceTask} of given ID from maintenance registry.
+     *
+     * @param mntcId
+     */
+    public void unregisterMaintenanceTask(UUID mntcId);
+
+    /**
+     * Returns active {@link MaintenanceTask} by its ID.
+     * There are active tasks only when node entered Maintenance Mode.
+     *
+     * {@link MaintenanceTask} becomes active when node enters Maintenance Mode and doesn't resolve the task
+     * during maintenance prepare phase.
+     *
+     * @return {@link MaintenanceTask} object for given maintenance ID or null if no maintenance task was found.
+     */
+    @Nullable public MaintenanceTask activeMaintenanceTask(UUID maitenanceId);
+
+    /**
+     * @param id UUID of {@link MaintenanceTask} this callback is registered for.
```

Comment:
Missing description

Reply:
fixed javadoc here and in other places in this class